File: server/internal/voice/turn.go
package voice

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha1"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"sync"
	"time"

	"github.com/google/uuid"
)

// TURNCredentialProvider generates ICE server credentials for WebRTC clients.
type TURNCredentialProvider interface {
	GetICEServers() (json.RawMessage, error)
}

// --- Cloudflare TURN (legacy) ---

// CFTurnConfig holds Cloudflare TURN API credentials.
type CFTurnConfig struct {
	KeyID    string
	APIToken string
}

// CFTurnClient fetches short-lived TURN credentials from Cloudflare.
type CFTurnClient struct {
	config CFTurnConfig
	client *http.Client

	mu          sync.RWMutex
	cached      *CFTurnResponse
	cachedUntil time.Time
}

// CFTurnResponse is the response from Cloudflare's generate-ice-servers API.
type CFTurnResponse struct {
	ICEServers json.RawMessage `json:"iceServers"`
}

func NewCFTurnClient(cfg CFTurnConfig) *CFTurnClient {
	return &CFTurnClient{
		config: cfg,
		client: &http.Client{Timeout: 10 * time.Second},
	}
}

// GetICEServers returns the iceServers array from Cloudflare, caching for 1 hour.
func (c *CFTurnClient) GetICEServers() (json.RawMessage, error) {
	c.mu.RLock()
	if c.cached != nil && time.Now().Before(c.cachedUntil) {
		result := c.cached.ICEServers
		c.mu.RUnlock()
		return result, nil
	}
	c.mu.RUnlock()

	c.mu.Lock()
	defer c.mu.Unlock()

	// Double-check after acquiring write lock
	if c.cached != nil && time.Now().Before(c.cachedUntil) {
		return c.cached.ICEServers, nil
	}

	url := fmt.Sprintf("https://rtc.live.cloudflare.com/v1/turn/keys/%s/credentials/generate-ice-servers", c.config.KeyID)
	body := []byte(`{"ttl": 86400}`)

	req, err := http.NewRequest("POST", url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+c.config.APIToken)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("cloudflare TURN API request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("cloudflare TURN API returned %d: %s", resp.StatusCode, string(respBody))
	}

	var result CFTurnResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decode CF TURN response: %w", err)
	}

	c.cached = &result
	c.cachedUntil = time.Now().Add(1 * time.Hour)

	slog.Debug("fetched fresh Cloudflare TURN credentials")
	return result.ICEServers, nil
}

// --- Self-hosted TURN ---

// SelfHostedTurnClient generates HMAC-SHA1 credentials for a self-hosted TURN server.
type SelfHostedTurnClient struct {
	sharedSecret string
	turnURLs     []string
	ttl          time.Duration
}

// NewSelfHostedTurnClient creates a credential provider for a self-hosted TURN server.
func NewSelfHostedTurnClient(sharedSecret string, turnURLs []string, ttl time.Duration) *SelfHostedTurnClient {
	return &SelfHostedTurnClient{
		sharedSecret: sharedSecret,
		turnURLs:     turnURLs,
		ttl:          ttl,
	}
}

// GetICEServers generates fresh HMAC-SHA1 credentials and returns an iceServers JSON array.
func (c *SelfHostedTurnClient) GetICEServers() (json.RawMessage, error) {
	expiry := time.Now().Add(c.ttl).Unix()
	username := fmt.Sprintf("%d:%s", expiry, uuid.New().String()[:8])

	mac := hmac.New(sha1.New, []byte(c.sharedSecret))
	mac.Write([]byte(username))
	password := base64.StdEncoding.EncodeToString(mac.Sum(nil))

	iceServer := struct {
		URLs       []string `json:"urls"`
		Username   string   `json:"username"`
		Credential string   `json:"credential"`
	}{
		URLs:       c.turnURLs,
		Username:   username,
		Credential: password,
	}

	data, err := json.Marshal([]interface{}{iceServer})
	if err != nil {
		return nil, fmt.Errorf("marshal ICE servers: %w", err)
	}

	return json.RawMessage(data), nil
}

File: server-rs/src/voice/turn.rs
use std::sync::Arc;
use std::time::{Duration, Instant};

use hmac::{Hmac, Mac};
use sha1::Sha1;
use tokio::sync::RwLock;

/// TURNCredentialProvider generates ICE server credentials for WebRTC clients.
#[async_trait::async_trait]
pub trait TURNCredentialProvider: Send + Sync {
    async fn get_ice_servers(&self) -> Result<serde_json::Value, String>;
}

// ---------------------------------------------------------------------------
// Cloudflare TURN
// ---------------------------------------------------------------------------

/// CFTurnConfig holds Cloudflare TURN API credentials.
#[derive(Clone)]
pub struct CFTurnConfig {
    pub key_id: String,
    pub api_token: String,
}

/// CFTurnClient fetches short-lived TURN credentials from Cloudflare.
pub struct CFTurnClient {
    config: CFTurnConfig,
    client: reqwest::Client,
    cache: Arc<RwLock<Option<CachedCredentials>>>,
}

struct CachedCredentials {
    ice_servers: serde_json::Value,
    valid_until: Instant,
}

impl CFTurnClient {
    pub fn new(config: CFTurnConfig) -> Self {
        let client = reqwest::Client::builder()
            .timeout(Duration::from_secs(10))
            .build()
            .unwrap_or_default();

        CFTurnClient {
            config,
            client,
            cache: Arc::new(RwLock::new(None)),
        }
    }
}

#[async_trait::async_trait]
impl TURNCredentialProvider for CFTurnClient {
    async fn get_ice_servers(&self) -> Result<serde_json::Value, String> {
        // Check cache (read lock).
        {
            let cache = self.cache.read().await;
            if let Some(ref cached) = *cache {
                if Instant::now() < cached.valid_until {
                    return Ok(cached.ice_servers.clone());
                }
            }
        }

        // Acquire write lock and double-check.
        let mut cache = self.cache.write().await;
        if let Some(ref cached) = *cache {
            if Instant::now() < cached.valid_until {
                return Ok(cached.ice_servers.clone());
            }
        }

        let url = format!(
            "https://rtc.live.cloudflare.com/v1/turn/keys/{}/credentials/generate-ice-servers",
            self.config.key_id
        );

        let body = serde_json::json!({"ttl": 86400});

        let resp = self
            .client
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.config.api_token))
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("cloudflare TURN API request failed: {e}"))?;

        let status = resp.status();
        if status != reqwest::StatusCode::CREATED && status != reqwest::StatusCode::OK {
            let resp_body = resp
                .text()
                .await
                .unwrap_or_else(|_| "<unreadable>".to_string());
            return Err(format!(
                "cloudflare TURN API returned {}: {}",
                status, resp_body
            ));
        }

        let result: serde_json::Value = resp
            .json()
            .await
            .map_err(|e| format!("decode CF TURN response: {e}"))?;

        let ice_servers = result
            .get("iceServers")
            .cloned()
            .ok_or_else(|| "missing iceServers in CF response".to_string())?;

        *cache = Some(CachedCredentials {
            ice_servers: ice_servers.clone(),
            valid_until: Instant::now() + Duration::from_secs(3600),
        });

        tracing::debug!("fetched fresh Cloudflare TURN credentials");
        Ok(ice_servers)
    }
}

// ---------------------------------------------------------------------------
// Self-hosted TURN (HMAC-SHA1 credentials)
// ---------------------------------------------------------------------------

/// SelfHostedTurnClient generates HMAC-SHA1 credentials for a self-hosted TURN server.
pub struct SelfHostedTurnClient {
    shared_secret: String,
    turn_urls: Vec<String>,
    ttl: Duration,
}

impl SelfHostedTurnClient {
    pub fn new(shared_secret: String, turn_urls: Vec<String>, ttl: Duration) -> Self {
        SelfHostedTurnClient {
            shared_secret,
            turn_urls,
            ttl,
        }
    }
}

#[async_trait::async_trait]
impl TURNCredentialProvider for SelfHostedTurnClient {
    async fn get_ice_servers(&self) -> Result<serde_json::Value, String> {
        let expiry = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
            + self.ttl.as_secs();

        let short_id = &uuid::Uuid::new_v4().to_string()[..8];
        let username = format!("{}:{}", expiry, short_id);

        let mut mac = Hmac::<Sha1>::new_from_slice(self.shared_secret.as_bytes())
            .map_err(|e| format!("HMAC key error: {e}"))?;
        mac.update(username.as_bytes());
        let password = base64::Engine::encode(
            &base64::engine::general_purpose::STANDARD,
            mac.finalize().into_bytes(),
        );

        let ice_server = serde_json::json!([{
            "urls": self.turn_urls,
            "username": username,
            "credential": password,
        }]);

        Ok(ice_server)
    }
}
